datafusion/expr/src/logical_plan/extension.rs

```diff
@@ -54,6 +57,22 @@ pub trait UserDefinedLogicalNode: fmt::Debug + Send + Sync {
     /// Return the output schema of this logical plan node.
     fn schema(&self) -> &DFSchemaRef;
 
+    /// Return the list of invariants.
+    ///
+    /// Implementing this function enables the user to define the
+    /// invariants for a given logical plan extension.
+    fn invariants(&self) -> Vec<Invariant> {
+        vec![]
+    }
+
+    /// Perform check of invariants for the extension node.
+    fn check_invariants(&self, check: InvariantLevel, plan: &LogicalPlan) -> Result<()> {
```

Comment:
I didn't understand the need to have both 

```rust
    fn invariants(&self) -> Vec<Invariant> {
```

And 
```
    fn check_invariants(&self, check: InvariantLevel, plan: &LogicalPlan) -> Result<()> {
```

If there is a reason to have both, can you please add to the documentation guidance for when the user would define one or the other?  

If there is no reason for both, perhaps we can start with just adding `check_invariants` (and remove `invariants` from thi PR)  which is the more flexible API

Reply:
Yeah, having both only makes sense when considering the next step of having a user define an `Invariant` which can be run on any LogicalPlan node. The user provides this abstraction:
```
pub struct Invariant {
    pub kind: InvariantLevel,
    pub fun: InvariantFn,
}
```

for either for a their own extension node. Or to the session context to run on any builtin Logical Plan node.

However, if we are not going to do this follow up bit -- then yes, I can consolidate it to only the single `fn check_invariants()` method.

Reply:
> However, if we are not going to do this follow up bit -- then yes

I reocmmend this be the last PR for a while. My rationale is 
1. you have added a nice API but there aren't many uses of it in the code base yet and thus we are speculating (likely correctly) on the value of additional APIs
2. Users can implement checks for existing LogicalPlans using existing APIs (not as nicely, but they could write their own pass)

So let's wrap this one up, and then focus on using what we have to make the code better